src/openvic-simulation/map/Province.cpp

```diff
@@ -116,57 +124,241 @@ void Province::tick(Date today) {
 	}
 }
 
-Province::adjacency_t::adjacency_t(Province const* province, distance_t distance, flags_t flags)
-	: province { province }, distance { distance }, flags { flags } {
-	assert(province != nullptr);
+Province::adjacency_t::adjacency_t(
+	Province const* new_to, distance_t new_distance, type_t new_type, Province const* new_through, data_t new_data
+) : to { new_to }, distance { new_distance }, type { new_type }, through { new_through }, data { new_data } {}
+
+std::string_view Province::adjacency_t::get_type_name(type_t type) {
+	switch (type) {
+	case type_t::LAND: return "Land";
+	case type_t::WATER: return "Water";
+	case type_t::COASTAL: return "Coastal";
+	case type_t::IMPASSABLE: return "Impassable";
+	case type_t::STRAIT: return "Strait";
+	case type_t::CANAL: return "Canal";
+	default: return "Invalid Adjacency Type";
+	}
+}
+
+Province::adjacency_t* Province::get_adjacency_to(Province const* province) {
+	const std::vector<adjacency_t>::iterator it = std::find_if(adjacencies.begin(), adjacencies.end(),
+		[province](adjacency_t const& adj) -> bool { return adj.get_to() == province; }
+	);
+	if (it != adjacencies.end()) {
+		return &*it;
+	} else {
+		return nullptr;
+	}
+}
+
+Province::adjacency_t const* Province::get_adjacency_to(Province const* province) const {
+	const std::vector<adjacency_t>::const_iterator it = std::find_if(adjacencies.begin(), adjacencies.end(),
+		[province](adjacency_t const& adj) -> bool { return adj.get_to() == province; }
+	);
+	if (it != adjacencies.end()) {
+		return &*it;
+	} else {
+		return nullptr;
+	}
 }
 
 bool Province::is_adjacent_to(Province const* province) const {
-	for (adjacency_t adj : adjacencies) {
-		if (adj.province == province) {
-			return true;
+	return province != nullptr && std::any_of(adjacencies.begin(), adjacencies.end(),
+		[province](adjacency_t const& adj) -> bool { return adj.get_to() == province; }
+	);
+}
+
+std::vector<Province::adjacency_t const*> Province::get_adjacencies_going_through(Province const* province) const {
+	std::vector<adjacency_t const*> ret;
+	for (adjacency_t const& adj : adjacencies) {
+		if (adj.get_through() == province) {
+			ret.push_back(&adj);
 		}
 	}
-	return false;
+	return ret;
+}
+
+bool Province::has_adjacency_going_through(Province const* province) const {
+	return province != nullptr && std::any_of(adjacencies.begin(), adjacencies.end(),
+		[province](adjacency_t const& adj) -> bool { return adj.get_through() == province; }
+	);
 }
 
-bool Province::add_adjacency(Province const* province, distance_t distance, flags_t flags) {
-	if (province == nullptr) {
-		Logger::error("Tried to create null adjacency province for province ", get_identifier(), "!");
+/* This is called for all adjacent pixel pairs and returns whether or not a new adjacency was add,
+ * hence the lack of error messages in the false return cases. */
+bool Province::add_standard_adjacency(Province& from, Province& to) {
```

Comment:
Nice refactor. I though about this in my previous PR but decided it was out of scope there.
It fits well in this PR.